README.md
# postcss-lazyimagecss 

<img align="right" width="130" height="130" title="PostCSS" src="http://postcss.github.io/postcss/logo.svg">

[![Build Status](https://travis-ci.org/Jeff2Ma/postcss-lazyimagecss.svg?branch=master)](https://travis-ci.org/Jeff2Ma/postcss-lazyimagecss)
[![npm version](https://badge.fury.io/js/postcss-lazyimagecss.svg)](http://badge.fury.io/js/postcss-lazyimagecss)

A [PostCSS](https://github.com/postcss/postcss) plugin that generates images's CSS `width` & `height` properties from stylesheets automatically.

Another lazy way to write CSS, feel free to use it :)

Based on [gulp-lazyimagecss](https://github.com/weixin/gulp-lazyimagecss). Thanks to [hzlzh](https://github.com/hzlzh) and [littledu](https://github.com/littledu).

```css
/* Input */
.icon-close {
	background-image: url(../slice/icon-close.png); //icon-close.png - 16x16
}

.icon-new {
	background-image: url(../slice/icon-new@2x.png); //icon-new@2x.png - 16x16
}
	
/* Output */
.icon-close {
	background-image: url(../slice/icon-close.png);
	width: 16px;
	height: 16px;
}

.icon-new {
	background-image: url(../slice/icon-new@2x.png);
	width: 8px;
	height: 8px;
	background-size: 8px 8px;
}

```

## Features

- Support `jpg`/`jpeg`/`png`/`gif`/`bmp`/`svg` file type.

- Support retina image (file name should like `demo@2x.png`).

- Both `background-image: url()` and `background: url()` can be detected successfully.

- CSS property generating will be ignored if any of those `width` / `height` / `background-size` already set.


## Installation

Install with npm:

	npm install postcss-lazyimagecss --save-dev
	

Or install width [yarn](https://github.com/yarnpkg/yarn):

	yarn add postcss-lazyimagecss --dev

## Usage

### Work with [Gulp](http://gulpjs.com/)

Example:

```js
var gulp = require('gulp');
var postcss = require('gulp-postcss');
var lazyimagecss = require('postcss-lazyimagecss');

gulp.task('css', function () {
	return gulp.src('./src/css/*.css')
	    .pipe(another-plugin())
		.pipe(postcss([lazyimagecss({
			imagePath: ['../img','../slice']
		})]))
		.pipe(gulp.dest('./dist/css'));
});
```	

### Work with Gulp & [gulp-sourcemaps](https://www.npmjs.com/package/gulp-sourcemaps)

Example:

```js
var gulp = require('gulp');
var postcss = require('gulp-postcss');
var lazyimagecss = require('postcss-lazyimagecss');
var sourcemaps = require('gulp-sourcemaps');

gulp.task('css', function () {
	return gulp.src('./src/css/*.css')
		.pipe(sourcemaps.init())
		.pipe(another-plugin())
		.pipe(postcss([lazyimagecss({
			imagePath: ['../img','../slice']
		})]))
		.pipe(sourcemaps.write("."))
		.pipe(gulp.dest('./dist'));
});
```

## Options
- **imagePath** Set image path to be worked (e.g. `['../slice','../img']`)

- **width**  Whether output `width` properties in CSS ( default: `true` )

- **height**  Whether output `height` properties in CSS ( default: `true` )

- **backgroundSize** Whether output `background-size` properties in CSS ( default: `true` )

## Contributing

[Issues](https://github.com/Jeff2Ma/postcss-lazyimagecss/issues/) and [Pull requests](https://github.com/Jeff2Ma/postcss-lazyimagecss/pulls) are welcome.

```shell
$ git clone https://github.com/Jeff2Ma/postcss-lazyimagecss
$ cd postcss-lazyimagecss
$ npm i
$ gulp
```


CHANGELOG.md

# Change Log

Change log for postcss-lazyimagecss.

## [0.1.3] - 2016-12-21

- Add support for detecting even dimensions.

## [0.1.2] - 2016-11-07

- Fix log grammar.

## [0.1.1] - 2016-11-05

- Fix image file or a directory path bugs.

- Add `chalk` package for a better console log.

## [0.1.0] - 2016-11-02

- Fix bug and release to postcss official github page.

## [0.0.11] - 2016-10-15

- Fix bug with multi path in background-image url.

## [0.0.10] - 2016-10-10

- Add support for multi path in background-image url.

## [0.0.8] - 2016-10-09

- Initial release.


test/test.js
var should = require('should');
var vfs = require('vinyl-fs');
var through2 = require('through2');
var postcss = require('gulp-postcss');
var noop = function () {};
var lazyimagecss = require('../index.js');

describe('postcss-lazyimagecss Unit est', function() {
	it('Image `width` -> should be able to get `width` properties.', function(done) {
		vfs.src('./test/src/css/style.css')
			.pipe(postcss([lazyimagecss({
				imagePath: ['../img','../slice']
			})]))
			.pipe(through2.obj(function(file, enc, cb){
				content = file.contents.toString();
				content.match(/width/g).length.should.equal(14);
				cb();
			}))
			.on('data', noop)
			.on('end', done);
	});

	it('Image `height` -> should be able to get `height` properties.', function(done) {
		vfs.src('./test/src/css/style.css')
			.pipe(postcss([lazyimagecss({
				imagePath: ['../img','../slice']
			})]))
			.pipe(through2.obj(function(file, enc, cb){
				content = file.contents.toString();
				content.match(/height/g).length.should.equal(12);
				cb();
			}))
			.on('data', noop)
			.on('end', done);
	});

	it('Image `background-size` -> should be able to get `background-size` properties.', function(done) {
	vfs.src('./test/src/css/style.css')
		.pipe(postcss([lazyimagecss({
			imagePath: ['../img','../slice']
		})]))
		.pipe(through2.obj(function(file, enc, cb){
			content = file.contents.toString();
			content.match(/background-size/g).length.should.equal(7);
			cb();
		}))
		.on('data', noop)
		.on('end', done);
	});

	it('`width` value -> should be able to get correct width value.', function(done) {
		vfs.src('./test/src/css/style.css')
			.pipe(postcss([lazyimagecss({
				imagePath: ['../img','../slice']
			})]))
			.pipe(through2.obj(function(file, enc, cb){
				content = file.contents.toString();
				content.indexOf('width: 14px').should.be.above(0);
				cb();
			}))
			.on('data', noop)
			.on('end', done);
	});

	it('`height` value -> should be able to get correct height value.', function(done) {
		vfs.src('./test/src/css/style.css')
			.pipe(postcss([lazyimagecss({
				imagePath: ['../img','../slice']
			})]))
			.pipe(through2.obj(function(file, enc, cb){
				content = file.contents.toString();
				content.indexOf('height: 16px').should.be.above(2);
				cb();
			}))
			.on('data', noop)
			.on('end', done);
	});

	it('`background-size` value -> should be able to get correct height value.', function(done) {
		vfs.src('./test/src/css/style.css')
			.pipe(postcss([lazyimagecss({
				imagePath: ['../img','../slice']
			})]))
			.pipe(through2.obj(function(file, enc, cb){
				content = file.contents.toString();
				content.indexOf('background-size: 14px 14px').should.be.above(0);
				cb();
			}))
			.on('data', noop)
			.on('end', done);
	});


	it(' Option: imagePath -> should work with another Option of imagePath.', function(done) {
		vfs.src('./test/src/css/style.css')
			.pipe(postcss([lazyimagecss({
				imagePath: ['../slice']
			})]))
			.pipe(through2.obj(function(file, enc, cb){
				content = file.contents.toString();
				content.match(/width/g).length.should.equal(12);
				content.match(/height/g).length.should.equal(10);
				content.match(/background-size/g).length.should.equal(7);
				cb();
			}))
			.on('data', noop)
			.on('end', done);
	});

	it(' Option: width -> should work when disable Option width.', function(done) {
		vfs.src('./test/src/css/style.css')
			.pipe(postcss([lazyimagecss({
				imagePath: ['../slice'],
				width: false
			})]))
			.pipe(through2.obj(function(file, enc, cb){
				content = file.contents.toString();
				content.match(/width/g).length.should.equal(4);
				cb();
			}))
			.on('data', noop)
			.on('end', done);
	});

	it(' Option: height -> should work when disable Option height.', function(done) {
		vfs.src('./test/src/css/style.css')
			.pipe(postcss([lazyimagecss({
				imagePath: ['../slice'],
				height: false
			})]))
			.pipe(through2.obj(function(file, enc, cb){
				content = file.contents.toString();
				content.match(/height/g).length.should.equal(1);
				cb();
			}))
			.on('data', noop)
			.on('end', done);
	});

	it(' Option: backgroundSize -> should work when disable Option backgroundSize.', function(done) {
		vfs.src('./test/src/css/style.css')
			.pipe(postcss([lazyimagecss({
				imagePath: ['../slice'],
				backgroundSize: false
			})]))
			.pipe(through2.obj(function(file, enc, cb){
				content = file.contents.toString();
				content.match(/background-size/g).length.should.equal(3);
				cb();
			}))
			.on('data', noop)
			.on('end', done);
	});
});



index.js
var path = require('path');
var fs = require('fs');
var dirname = require('path').dirname;
var postcss = require('postcss');
var chalk = require('chalk');
var _ = require('lodash');
var fastImageSize = require('./lib/fastimagesize');

/**
 * a helper to find the real image absolute path,
 * deal with the issue like `../../img.jpg` and so on.
 */
function fixAbsolutePath(dir, relative) {
	// find the first time
	var absolute = path.resolve(dir, relative);

	// check if is a image file
	var reg = /\.(jpg|jpeg|png|gif|svg|bmp)\b/i;
	if (!reg.test(absolute)) {
		pluginLog('Not a image file: ', absolute);
		return;
	}

	if (!fs.existsSync(absolute) && (relative.indexOf('../') > -1)) {
		relative = relative.replace('../', '');
		// find the second time
		absolute = path.resolve(dir, relative);
	}

	return absolute;
}

function pluginLog(str, arg) {
	return console.log('[' + chalk.blue('postcss-lazyimagecss') + '] ' + chalk.red(str) + arg);
}

/**
 * main function
 */
module.exports = postcss.plugin('lazyimagecss', function (options) {
	return function (css) {
		options = options || {};

		options = _.extend({
			width: true,
			height: true,
			backgroundSize: true,
			imagePath: []
		}, options);

		var imagePath = options.imagePath;

		if (imagePath.length) {
			imagePath = '(' + options.imagePath.join('|') + '/)';
			imagePath = imagePath.replace(/\./g, '\\.');
		} else {
			imagePath = '';
		}

		var imageRegex = new RegExp('url\\(["\']?(' + imagePath + '[^)]*?)["\']?\\)');

		css.walkRules(function (rule) {
			rule.walkDecls(/^background(-image)?$/, function (decl) {
				var rule = decl.parent;
				var nodes = rule.nodes;
				var value = decl.value;
				// var prop = decl.prop;
				var CSSWidth = false;
				var CSSHeight = false;
				var CSSBGSize = false;

				var matchValue = imageRegex.exec(value);

				if (!matchValue || matchValue[1].indexOf('data:') === 0) {
					return;
				}

				var relativePath = matchValue[1];

				var inputDir = dirname(decl.source.input.file);

				var absolutePath = fixAbsolutePath(inputDir, relativePath);

				if (absolutePath === undefined) {
					return;
				}

				nodes.forEach(function (node) {
					if (node.prop === 'width') {
						CSSWidth = true;
					}
					if (node.prop === 'height') {
						CSSHeight = true;
					}
					if (node.prop === 'background-size' || node.prop === '-webkit-background-size') {
						CSSBGSize = true;
					}
				});

				if (value.indexOf('@2x') > -1) {
					options.retina = true;
				} else {
					options.retina = false;
				}

				var info = fastImageSize(absolutePath);

				if (info === undefined) {
					pluginLog('File does not exist: ', absolutePath);
					return;
				}

				if (info.type === 'unknown') {
					pluginLog('Unknown type: ', absolutePath);
					return;
				}

				// check if even dimensions
				if (value.indexOf('@2x') > -1 && (info.width % 2 !== 0 || info.height % 2 !== 0)) {
					pluginLog('Should have even dimensions: ', absolutePath);
					return;
				}

				var valueWidth, valueHeight;

				if (options.retina) {
					valueWidth = (info.width / 2) + 'px';
					valueHeight = (info.height / 2) + 'px';
				} else {
					valueWidth = info.width + 'px';
					valueHeight = info.height + 'px';
				}

				if (options.width && !CSSWidth) {
					rule.append({prop: 'width', value: valueWidth});
				}

				if (options.height && !CSSHeight) {
					rule.append({prop: 'height', value: valueHeight});
				}

				if (options.backgroundSize && options.retina && !CSSBGSize) {
					rule.append({prop: 'background-size', value: valueWidth + ' ' + valueHeight});
				}
			});
		});
	};
});


gulpfile.js
var gulp = require('gulp');
var postcss = require('gulp-postcss');
var lazyimagecss = require('./index.js');
var mocha = require('gulp-mocha');

var files = ['index.js'];
var watchFiles = ['index.js', 'test/**/*'];

gulp.task('lint', function () {
	var eslint = require('gulp-eslint');
	return gulp.src(files)
		.pipe(eslint())
		.pipe(eslint.format())
		.pipe(eslint.failAfterError());
});

gulp.task('test', function () {
	return gulp.src('test/*.js', { read: false })
		.pipe(mocha({ timeout: 1000000 }));
});

gulp.task('css', function () {
	return gulp.src('./test/src/css/**/*.css')
		.pipe(postcss([lazyimagecss({
			imagePath: ['../img', '../slice']
		})]))
		.pipe(gulp.dest('./test/dist/css'));
});

gulp.task('default', ['watch']);

gulp.task('watch', function () {
	gulp.watch(watchFiles, ['css', 'test', 'lint']);
});


lib/fastimagesize.js
//
// fast-image-size - Simple stand alone module to just extract the image size from image file without using special image libraries.
//
// Please refer to README.md for this module's documentations.
//
// NOTE:
// - Before changing this code please refer to the 'hacking the code section' on README.md.
//
// Copyright (c) 2013 Ziv Barber;
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the
// 'Software'), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so, subject to
// the following conditions:
//
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

var fs = require('fs');
var path = require('path');

module.exports = exports = function (file_path, callback) {

    function getJpgSize(buffer_data, retInfo) {
        // Skip 5 chars, they are for signature
        buffer_data = buffer_data.slice(4);

        var i, next;
        while (buffer_data.length) {
            // read length of the next block
            i = buffer_data.readUInt16BE(0);

            // 0xFFC0 is baseline(SOF)
            // 0xFFC2 is progressive(SOF2)
            next = buffer_data[i + 1];
            if (next === 0xC0 || next === 0xC2) {
                return {
                    'height': buffer_data.readUInt16BE(i + 5),
                    'width': buffer_data.readUInt16BE(i + 7)
                };
            }

            // move to the next block
            buffer_data = buffer_data.slice(i + 2);
        }
    }

    function parseHeaderData(buffer_data, callback_data) {
        var retInfo = {};

        // Detect GIF:
        if (buffer_data[0] == 0x47 && buffer_data[1] == 0x49 && buffer_data[2] == 0x46) {
            retInfo.type = 'gif';
            retInfo.width = (buffer_data[7] * 256) + buffer_data[6];
            retInfo.height = (buffer_data[9] * 256) + buffer_data[8];

            // Detect JPEG:
        } else if (buffer_data[0] == 0xFF && buffer_data[1] == 0xD8 && buffer_data[2] == 0xFF && (buffer_data[3] == 0xE0 || buffer_data[3] == 0xE1)) {
            retInfo.type = 'jpeg';
            var size = getJpgSize(buffer_data, retInfo);
            retInfo.width = size.width;
            retInfo.height = size.height;

            // Detect PNG:
        } else if (buffer_data[0] == 137 && buffer_data[1] == 80 && buffer_data[2] == 78 && buffer_data[3] == 71 && buffer_data[4] == 13 && buffer_data[5] == 10 && buffer_data[6] == 26 && buffer_data[7] == 10) {
            retInfo.type = 'png';

            if (buffer_data[12] == 0x49 && buffer_data[13] == 0x48 && buffer_data[14] == 0x44 && buffer_data[15] == 0x52) {
                retInfo.width = (buffer_data[16] * 256 * 256 * 256) + (buffer_data[17] * 256 * 256) + (buffer_data[18] * 256) + buffer_data[19];
                retInfo.height = (buffer_data[20] * 256 * 256 * 256) + (buffer_data[21] * 256 * 256) + (buffer_data[22] * 256) + buffer_data[23];
            } // Endif.

            // Detect BMP:
        } else if (buffer_data[0] == 0x42 && buffer_data[1] == 0x4D) {
            retInfo.type = 'bmp';
            retInfo.width = (buffer_data[21] * 256 * 256 * 256) + (buffer_data[20] * 256 * 256) + (buffer_data[19] * 256) + buffer_data[18];
            retInfo.height = (buffer_data[25] * 256 * 256 * 256) + (buffer_data[24] * 256 * 256) + (buffer_data[23] * 256) + buffer_data[22];
        } // Endif.

        retInfo.image = file_path;
        if (!retInfo.type) {
            retInfo.type = 'unknown';
        } // Endif.

        if (callback_data) {
            callback_data(retInfo);
        } // Endif.

        return retInfo;
    };

    function getSVGImageSize(data) {
        var retInfo = {};

        retInfo.type = 'svg';
        retInfo.width = data.match(/<svg(.*?)width="(.*?)px/)[2];
        retInfo.height = data.match(/<svg(.*?)height="(.*?)px/)[2];
        return retInfo;
    }

    // Async mode:
    if (callback) {
        fs.exists(file_path, function (exists) {
            if (exists) {
                fs.open(file_path, "r", function (error, fd) {
                    var bufferSize = fs.fstatSync(fd).size;
                    var buffer = new Buffer(bufferSize);
                    fs.read(fd, buffer, 0, bufferSize, null, function (error, bytesRead, buffer) {
                        fs.close(fd);
                        parseHeaderData(buffer, callback);
                    });
                });


            } else {
                throw 'Error: Invalid file name.';
            } // Endif.
        });

        // Sync mode:
    } else {

        if(!fs.existsSync(file_path)){
            // console.log(file_path + ' not exist');
            return;
        }

        if(file_path.indexOf('.svg') > -1){
            var fd = fs.openSync(file_path, "r");
            var bufferSize = fs.fstatSync(fd).size;
            var buffer = new Buffer(bufferSize);
            var bytesRead = fs.readSync(fd, buffer, 0, bufferSize, 0);
            fs.closeSync(fd);
            return getSVGImageSize(buffer.toString(),null);
        }else{
            var fd = fs.openSync(file_path, "r");
            var bufferSize = fs.fstatSync(fd).size;
            var buffer = new Buffer(bufferSize);
            var bytesRead = fs.readSync(fd, buffer, 0, bufferSize, 0);
            fs.closeSync(fd);
            return parseHeaderData(buffer, null);
        }


    } // Endif.
};

